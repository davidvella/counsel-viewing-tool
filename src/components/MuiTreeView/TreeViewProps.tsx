export interface Tree {
  value: string;
  nodes?: Array<string | Tree>;
  id: string;
}

export interface Leaf { 
  value: string; 
  parent: Tree; 
  id: string; 
}


export interface MuiTreeViewProps {
  /**
   * The data to render as a tree view
   */
  tree: Tree[];
  /**
   * Callback function fired when a tree leaf is clicked.
   */
  onLeafClick?: (
    leaf: Leaf
  ) => void;
  /**
   * 
   */
  defaultCollapseIcon?: React.ReactNode;
  /**
   * 
   */
  defaultEndIcon?: React.ReactNode;
  /**
   * 
   */
  defaultExpandIcon?: React.ReactNode;
  /**
   * 
   */
  defaultParentIcon?: React.ReactNode;
}