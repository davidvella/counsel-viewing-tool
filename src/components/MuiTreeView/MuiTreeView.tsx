import React from 'react';
import { MuiTreeViewProps, Leaf } from './TreeViewProps';
import TreeNode from '../TreeNode/TreeNode';
import TreeView from '../TreeView/TreeView';
import memoize from 'fast-memoize';

/**
 * A React tree view for material-ui.
 */
export default class MuiTreeView extends React.Component<MuiTreeViewProps, any> {

    createMemoizeTree = memoize(
        (tree) => (tree),
        {
            serializer: ([tree]) =>
                `${JSON.stringify(tree)}`,
        }
    );


    handleLeafClick = (leaf: Leaf) => {
        if (this.props.onLeafClick) {
            this.props.onLeafClick(leaf);
        }
    };

    isLeaf(node: any) {
        return typeof node === 'string' || !node.nodes || !node.nodes.length;
    }

    getNodeValue(node: any) {
        return typeof node === 'string' ? node : node.value;
    }

    getNodeId(node: any) {
        if (typeof node === 'object') {
            return node.id;
        }
    }

    renderNode = ({ node, parent, depth = 0 }: { node: any, parent: any, depth: number }) => {
        const value = this.getNodeValue(node);
        const id = this.getNodeId(node);
        const isLeaf = this.isLeaf(node);

        if (isLeaf) {
            return (
                <TreeNode nodeId={id} label={value} onClick={() => this.handleLeafClick({ ...node, value, parent, id })} key={typeof node.id !== 'undefined' ? node.id : node.value} />
            );
        }

        return (
            <TreeNode nodeId={id} label={value} key={typeof node.id !== 'undefined' ? node.id : node.value}>
                {node.nodes.map((l: any) =>
                    this.renderNode({
                        node: l,
                        parent: node,
                        depth: depth + 1
                    })
                )}
            </TreeNode>
        );
    };

    /**
     *  Default React component render method
     */
    public render() {
        const { tree,
            defaultCollapseIcon,
            defaultEndIcon,
            defaultExpandIcon,
            defaultParentIcon } = this.props;
        const graph = this.createMemoizeTree(tree);

        return (
            <TreeView defaultCollapseIcon={defaultCollapseIcon} defaultExpandIcon={defaultExpandIcon} defaultEndIcon={defaultEndIcon} defaultParentIcon={defaultParentIcon}>
                {
                    graph.map((node: any) =>
                        this.renderNode({ node, parent: null, depth: 0 })
                    )
                }
            </TreeView>
        );
    }

}