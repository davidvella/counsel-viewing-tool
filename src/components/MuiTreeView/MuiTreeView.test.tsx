import React from 'react';
import {
  render
} from "@testing-library/react";
// this adds custom jest matchers from jest-dom
import "jest-dom/extend-expect";
import MuiTreeView from "./MuiTreeView";
import { Tree } from "./TreeViewProps";


const testTree: Tree[] = [
  {
    value: 'Parent A',
    nodes: [{ value: 'Child A', id: "2" }, { value: 'Child B', id: "3" }],
    id: "1"
  },
  {
    value: 'Parent B',
    id: "4",
    nodes: [
      {
        id: "5",
        value: 'Child C',
      },
      {
        id: "6",
        value: 'Parent C',
        nodes: [
          { value: 'Child D', id: "7" },
          { value: 'Child E', id: "8" },
          {
            value: 'Child F', id: "9", nodes: [
              {
                id: "10",
                value: 'Child C',
                nodes: [
                  {
                    id: "11",
                    value: 'Child C',
                  }
                ]
              }
            ]
          },
        ],
      },
    ],
  },
];

describe('<MuiTreeView /> spec', () => {

  it("renders the component", () => {
    const { container } = render(<MuiTreeView tree={testTree} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('assert there are 11 leafs', () => {
    expect(document.querySelectorAll('.MuiTypography-body1').length).toBe(11)
  })

  it('assert the first leaf to be Parent A', () => {
    const leaf = document.querySelector('.MuiTypography-body1:first-child')
    expect(leaf).toBeTruthy();
    expect(leaf!.innerHTML).toBe('Parent A')
   });

});

