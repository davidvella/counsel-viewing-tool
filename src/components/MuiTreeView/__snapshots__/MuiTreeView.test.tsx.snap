// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MuiTreeView /> spec renders the component 1`] = `
<ul
  class="ForwardRef(TreeView)-root-1"
  role="tree"
>
  <li
    aria-expanded="false"
    class="ForwardRef(TreeNode)-root-2"
    role="treeitem"
    tabindex="0"
  >
    <div
      class="ForwardRef(TreeNode)-content-4"
    >
      <p
        class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
      >
        Parent A
      </p>
    </div>
    <ul
      class="MuiCollapse-container ForwardRef(TreeNode)-group-3 MuiCollapse-hidden"
      role="group"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <li
            class="ForwardRef(TreeNode)-root-2"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="ForwardRef(TreeNode)-content-4"
            >
              <p
                class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
              >
                Child A
              </p>
            </div>
          </li>
          <li
            class="ForwardRef(TreeNode)-root-2"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="ForwardRef(TreeNode)-content-4"
            >
              <p
                class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
              >
                Child B
              </p>
            </div>
          </li>
        </div>
      </div>
    </ul>
  </li>
  <li
    aria-expanded="false"
    class="ForwardRef(TreeNode)-root-2"
    role="treeitem"
    tabindex="-1"
  >
    <div
      class="ForwardRef(TreeNode)-content-4"
    >
      <p
        class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
      >
        Parent B
      </p>
    </div>
    <ul
      class="MuiCollapse-container ForwardRef(TreeNode)-group-3 MuiCollapse-hidden"
      role="group"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <li
            class="ForwardRef(TreeNode)-root-2"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="ForwardRef(TreeNode)-content-4"
            >
              <p
                class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
              >
                Child C
              </p>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="ForwardRef(TreeNode)-root-2"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="ForwardRef(TreeNode)-content-4"
            >
              <p
                class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
              >
                Parent C
              </p>
            </div>
            <ul
              class="MuiCollapse-container ForwardRef(TreeNode)-group-3 MuiCollapse-hidden"
              role="group"
              style="min-height: 0px;"
            >
              <div
                class="MuiCollapse-wrapper"
              >
                <div
                  class="MuiCollapse-wrapperInner"
                >
                  <li
                    class="ForwardRef(TreeNode)-root-2"
                    role="treeitem"
                    tabindex="-1"
                  >
                    <div
                      class="ForwardRef(TreeNode)-content-4"
                    >
                      <p
                        class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
                      >
                        Child D
                      </p>
                    </div>
                  </li>
                  <li
                    class="ForwardRef(TreeNode)-root-2"
                    role="treeitem"
                    tabindex="-1"
                  >
                    <div
                      class="ForwardRef(TreeNode)-content-4"
                    >
                      <p
                        class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
                      >
                        Child E
                      </p>
                    </div>
                  </li>
                  <li
                    aria-expanded="false"
                    class="ForwardRef(TreeNode)-root-2"
                    role="treeitem"
                    tabindex="-1"
                  >
                    <div
                      class="ForwardRef(TreeNode)-content-4"
                    >
                      <p
                        class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
                      >
                        Child F
                      </p>
                    </div>
                    <ul
                      class="MuiCollapse-container ForwardRef(TreeNode)-group-3 MuiCollapse-hidden"
                      role="group"
                      style="min-height: 0px;"
                    >
                      <div
                        class="MuiCollapse-wrapper"
                      >
                        <div
                          class="MuiCollapse-wrapperInner"
                        >
                          <li
                            aria-expanded="false"
                            class="ForwardRef(TreeNode)-root-2"
                            role="treeitem"
                            tabindex="-1"
                          >
                            <div
                              class="ForwardRef(TreeNode)-content-4"
                            >
                              <p
                                class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
                              >
                                Child C
                              </p>
                            </div>
                            <ul
                              class="MuiCollapse-container ForwardRef(TreeNode)-group-3 MuiCollapse-hidden"
                              role="group"
                              style="min-height: 0px;"
                            >
                              <div
                                class="MuiCollapse-wrapper"
                              >
                                <div
                                  class="MuiCollapse-wrapperInner"
                                >
                                  <li
                                    class="ForwardRef(TreeNode)-root-2"
                                    role="treeitem"
                                    tabindex="-1"
                                  >
                                    <div
                                      class="ForwardRef(TreeNode)-content-4"
                                    >
                                      <p
                                        class="MuiTypography-root ForwardRef(TreeNode)-label-6 MuiTypography-body1"
                                      >
                                        Child C
                                      </p>
                                    </div>
                                  </li>
                                </div>
                              </div>
                            </ul>
                          </li>
                        </div>
                      </div>
                    </ul>
                  </li>
                </div>
              </div>
            </ul>
          </li>
        </div>
      </div>
    </ul>
  </li>
</ul>
`;
